File: examples/python/good/ocp.py
import abc


class BaseReportSender(abc.ABC):
    def __init__(self, report):
        self._report = report

    @abc.abstractmethod
    def send(self, to):
        pass


class EmailSender(BaseReportSender):
    def __init__(self, report):
        super().__init__(report, from_email)
        self._from = from_email

    def send(self, to):
        print(f"send email from: {self._from} to: {to}")


class MSTeamsSender(BaseReportSender):
    def __init__(self, report, api_key):
        super().__init__(report)
        self._api_key = api_key

    def send(self, to):
        print(f"send msteam to: {to} using apikey")



def get_sender(sender_type):
    if sender_type == "email":
        return EmailSender("report data", "admin@test.com")
    elif sender_type == "msteams":
        return MSTeamsSender("report data", "key-xxxxx")
    else:
        raise ValueError("Invalid Sender type")


sender = get_sender("email")
sender.send("john@test.com")
sender = get_sender("msteams")
sender.send("jane")

File: examples/typescript/good/ocp.ts
abstract class BaseReportSender {
  private report: string;
  constructor(report: string) {
    this.report = report;
  }
  abstract send(to: string): void;
}

class EmailSender extends BaseReportSender {
  private from: string;
  constructor(report: string, from: string) {
    super(report);
    this.from = from;
  }
  send(to: string): void {
    console.log(`send email from: ${this.from} to: ${to}`);
  }
}

class MSTeamsSender extends BaseReportSender {
  private apikey: string;
  constructor(report: string, apikey: string) {
    super(report);
    this.apikey = apikey;
  }
  send(to: string): void {
    console.log(`send email from: ${this.apikey} to: ${to}`);
  }
}

function get_sender(sender_type: string): BaseReportSender {
  switch (sender_type) {
    case "email":
      return new EmailSender("report data", "admin@test.com");
    case "msteams":
      return new MSTeamsSender("report data", "key-xxxxx");
    default:
      throw Error("Invalid Sender type");
  }
}

(function () {
  const sender1 = get_sender("email");
  sender1.send("john@test.com");
  const sender2 = get_sender("msteams");
  sender2.send("jane");
})();
